File: src/DFS/SurroundedRegions.java
package DFS;

// https://leetcode.com/problems/surrounded-regions/
// We search for invalid candidates (and exclude them) rather than search for valid candidates.
public class SurroundedRegions {
    private static final int[][] dirs = {
            {0, 1},
            {0, -1},
            {1, 0},
            {-1, 0}
    };

    public void solve(char[][] board) {
        if (board.length == 0)
            return;

        int m = board.length;
        int n = board[0].length;

        for (int i = 0; i < m; i++) {
            if (board[i][0] == 'O') {
                markNotSurrounded(i, 0, board);
            }

            if (board[i][n - 1] == 'O') {
                markNotSurrounded(i, n -1, board);
            }
        }

        for (int i = 0; i < n; i++) {
            if (board[0][i] == 'O') {
                markNotSurrounded(0, i, board);
            }

            if (board[m-1][i] == 'O') {
                markNotSurrounded(m-1, i, board);
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }

                if (board[i][j] == '*') {
                    board[i][j] = 'O';
                }
            }
        }

    }

    // Mark 'O' not surrounded by 'X' as '*'
    private void markNotSurrounded(int x, int y, char[][] board) {
        board[x][y] = '*';
        for(int[] dir: dirs) {
            int nextX = x + dir[0];
            int nextY = y + dir[1];

            if (nextX < board.length && nextX >= 0 && nextY < board[0].length && nextY >= 0 && board[nextX][nextY] == 'O') {
                markNotSurrounded(nextX, nextY, board);
            }
        }
    }

    public static void main(String[] args) {
        SurroundedRegions s = new SurroundedRegions();
        char[][] board = {
                {'X','X','X','X'},{'X','O','O','X'},{'X','X','O','X'},{'X','O','X','X'}
        };

        s.solve(board);
        System.out.println(board);
    }
}

File: py/essential/SurroundedRegions.py
# https://leetcode.com/problems/surrounded-regions/description/?envType=problem-list-v2&envId=plakya4j
# Neet

from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n = len(board), len(board[0])
        dirs = [
            (0, 1),
            (1, 0),
            (0, -1),
            (-1, 0),
        ]

        def markNotSurrounded(i: int, j: int):
            board[i][j] = "N"
            for d in dirs:
                if (
                    0 <= i + d[0] < m
                    and 0 <= j + d[1] < n
                    and board[i + d[0]][j + d[1]] == "O"
                ):
                    markNotSurrounded(i + d[0], j + d[1])

        # Iterate through the borders and mark all "O" as "N"
        for i in range(m):
            if board[i][0] == "O":
                markNotSurrounded(i, 0)
            if board[i][n - 1] == "O":
                markNotSurrounded(i, n - 1)

        for i in range(n):
            if board[0][i] == "O":
                markNotSurrounded(0, i)
            if board[m - 1][i] == "O":
                markNotSurrounded(m - 1, i)

        # then iterate through the board and flip all "O" to "X" (surrounded) and all "N" to "O" (not surrounded)
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "N":
                    board[i][j] = "O"


# Test case
s = Solution()
board = [
    ["X", "O", "X", "O", "X", "O"],
    ["O", "X", "O", "X", "O", "X"],
    ["X", "O", "X", "O", "X", "O"],
    ["O", "X", "O", "X", "O", "X"],
]
s.solve(board)
print(
    board
)  # Expected output: [['X', 'X', 'X', 'X'], ['X', 'X', 'X', 'X'], ['X', 'X', 'X', 'X'], ['X', 'O', 'X', 'X']]
